src/boot/efi/util.c

```diff
@@ -769,6 +770,108 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
         return EFI_SUCCESS;
 }
 
+#define QEMU_KERNEL_LOADER_FS_MEDIA_GUID \
+        { 0x1428f772, 0xb64a, 0x441e, {0xb8, 0xc3, 0x9e, 0xbd, 0xd7, 0xf8, 0x93, 0xc7 }}
+
+#define QEMU_BOOT_ORDER_GUID \
+        { 0x668f4529, 0x63d0, 0x4bb5, {0xb6, 0x5d, 0x6f, 0xbb, 0x9d, 0x36, 0xa4, 0x4a }}
+
+/* detect direct boot */
+bool is_qemu_direct_boot(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return false;
+
+        /* 'qemu -kernel systemd-bootx64.efi' */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_VENDOR_DP &&
+            memcmp(&dp->Guid, &(EFI_GUID)QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) == 0)
+                return true;
+
+        /* loaded from firmware volume (sd-boot added to ovmf) */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_PIWG_FW_VOL_DP)
+                return true;
+
+        return false;
+}
+
+/* return the length of the device path in bytes, excluding the end tag */
+static size_t device_path_length(EFI_DEVICE_PATH *dp)
+{
+        size_t len;
+
+        for (len = 0; !IsDevicePathEnd(dp); dp = NextDevicePathNode(dp))
```

Comment:
Not a matter, but in this case, I'd like to initialize `len` when declared:
```c
size_t len = 0;

for (; !IsDevicePathEnd(dp); dp = NextDevicePathNode(dp))
        len += ...;
```

Reply:
It's less work to just look for the end node and do `(uint8_t*) end_node - (uint8_t*) stat_node`, like `make_file_device_path` does.

Reply:
Doesn't make a big difference IMHO.  In both cases you'll need a loop to walk the device path to find the end node ...

Reply:
I'll insist to keep things consistent.